packages/backend/src/managers/podmanConnection.ts

```diff
@@ -0,0 +1,73 @@
+/**********************************************************************
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import { type ContainerProviderConnection, type RegisterContainerConnectionEvent, provider } from '@podman-desktop/api';
+
+type startupHandle = () => void;
+
+export class PodmanConnection {
+  #firstFound = false;
+  #connection: ContainerProviderConnection | undefined = undefined;
+
+  #toExecuteAtStartup: startupHandle[] = [];
+
+  init(): void {
+    // In case the extension has not yet registered, we listen for new registrations
+    // and retain the first started podman provider
+    const disposable = provider.onDidRegisterContainerConnection((e: RegisterContainerConnectionEvent) => {
```

Comment:
Why are we listening on connection registration andnot connection start/stop events ?

Reply:
For this feature, I only want to list containers started before Podman Desktop is started, and I want to be informed when the podman extension is registered (or the `listContainers` would return an empty list if we try to retrieve the list before). 

(note that there is a `registries/ContainerRegistry` which is watching at events on a specific container, which we may merge with this class at some point)

Reply:
I didn't test it so maybe i'm wrong but looking at the code it looks to me that it initializes the connection when the studio is activated if there is atleast one podman machine that is started and then it starts listening to if a machine gets registered.
So if i start my desktop having my podman machine stopped, the connection is not set as there is no podman machine running. If i start my podman machine in a second moment it does not trigger its initialization bc it only listen to machine registration. 
So if i start the playground or any other feature that requires to retrieve the active connection it fails, no? But i have my podman machine running fine, i just started it after my desktop was activated

Reply:
Yes, you are probably right, I'll look at supporting this scenario too. Thanks

Reply:
Finally, I prefer for now to remove this exposure and use of `connection` from the `PodmcnConnection` class, as it is not part of what is needed to fix the targeted issue, which is only related to getting a connection to the provider **at startup**.

I'll work at this refactoring as part of another PR, if necessary.

Reply:
I still think that this PR is not complete. You're just adopting the containers if the podman machine is running when you activate desktop or if a new machine is created. It does not cover the other use cases.

Reply:
I have made a few tests on Windows, and I cannot find another scenario where we could adopt a container other than the one already covered: the playground container being created with the AUTOREMOVE flag (equivalent to --rm), when the podman machine is stopped, the container is stopped and deleted, and when the machine is restarted, the container is not present anymore. So, there is no way we can find a playground container when starting a podman machine.

Do you have other scenarios in mind?

Reply:
I wanted to test it but i'm not sure what i have to do 😓 So when starting ai-studio with a running machine + playground 
 i should go to the playground page and find it running? Bc i still see the play action and if i click it generates a new container🤔 

My other scenario is, you are using the playground, turn off the machine bc you switch to another or whatever. ai-studio should be able to remove the adopted container as it is stopped now, then you turn the machine on again, the container is there stopped, yo start it, ai-studio should be able to adopt it again.

BTW maybe my assumption was wrong as i expected it to work differently. How should i test it?

Reply:
Here is a demo on Linux to see what you can test:

https://github.com/projectatomic/ai-studio/assets/9973512/e290e3c2-4fb4-427b-a333-c4a196ccdf66

> My other scenario is, you are using the playground, turn off the machine bc you switch to another or whatever. ai-studio should be able to remove the adopted container as it is stopped now, then you turn the machine on again, the container is there stopped, yo start it, ai-studio should be able to adopt it again.

For now, we are not checking when a machine is stopped, to mark playground as stopped. It would be an extension of the PR https://github.com/projectatomic/ai-studio/pull/194 (and not particularly part of this PR as it is a different logic).

As said in my previous message, when you stop a playground container, it is also deleted (because marked as AUTOREMOVE). So you will never have a playground container (not even stopped) when you start a podman machine.

Reply:
If you stop the machine the container will be there, not deleted and you can restart it once you restart the machine.

![image](https://github.com/projectatomic/ai-studio/assets/49404737/bdc2a701-f17e-429b-b3e0-177124b404a6)

Reply:
AFAIK, it is not the expected behaviour for a container marked as auto-removable. In Inspect of this container, is it marked as this?
```
  "HostConfig": {
    [...]
    "AutoRemove": true,
```

Reply:
Not that I don't want to support these stopped containers, but if we consider having stopped playground containers, this adds a third state for the playground container in the UI (paused in addition to not-running and running), which we would also have to manage in the UI (set the playground in pause, etc). That would be a bigger change than just detecting a stopped container is started again.

Reply:
Sorry for the slow response. Done it again now and i confirm that the Autoremove flag is there